### PR TITLE
Various memory allocation improvements.

### PR DESCRIPTION
**Pull request by petermattis**
Place a fixed size initial stack and the lval inside the parser struct
so that they aren't allocated separately. Place $$char inside the parser
struct to simplify the implementation of Lookahead() and avoid an
allocation of the closure.

### CODE REVIEW

**Comment by tschottdorf** [comment]
that's also going upstream, right?

**Comment by petermattis** [comment]
I'll try. I have no confidence in it being accepted.

File: yacc.go [hunk 1]
```diff
@@ -3331,18 +3323,17 @@ type $$Parser interface {
 }
 
 type $$ParserImpl struct {
-	lookahead func() int
+	lval  $$SymType
+	stack [16]$$SymType
```
Comment (bdarnell) [inline yacc.go]: You should keep the constant and just change `stacksize` to 16 (and maybe rename to `initialStackSize` since the original code seems to be under the impression that it's a limit rather than an initial value). A change that just changes the constant without making the other changes to `$$ParserImpl` might be more likely to be accepted upstream, if this is where a substantial portion of the performance gains come from.

**Comment by bdarnell** [comment]
LGTM

File: yacc.go (hunk 1 above)
Comment (tamird) [inline yacc.go]: it seems to me that the performance gain comes from reusing the slice, and not necessarily from making it smaller.

**Comment by tamird** [comment]
LGTM

File: yacc.go (hunk 1 above)
Comment (bdarnell) [inline yacc.go]: That would be my intuition as well but Peter mentioned at lunch that the size appeared to be a significant factor (more pointers for the GC to traverse).

Comment (petermattis) [inline yacc.go]: @tamird The array isn't reused. A new `ParserImpl` is allocated for every call to `Parse`.

@bdarnell Ok, I'll keep `stacksize` and rename to `initialStackSize`.

File: yacc.go [hunk 2]
```diff
@@ -3470,22 +3469,20 @@ func $$Parse($$lex $$Lexer) int {
 
 func ($$rcvr *$$ParserImpl) Parse($$lex $$Lexer) int {
 	var $$n int
-	var $$lval $$SymType
 	var $$VAL $$SymType
 	var $$Dollar []$$SymType
 	_ = $$Dollar // silence set and not used
-	$$S := make([]$$SymType, $$MaxDepth)
+	$$S := $$rcvr.stack[:]
```
Comment (tamird) [inline yacc.go]: not sure why this change; if `ParserImpl` is never reused, it seems equivalent to not have made this change at all.

File: yacc.go [hunk 3]
```diff
@@ -3470,22 +3469,20 @@ func $$Parse($$lex $$Lexer) int {
 
 func ($$rcvr *$$ParserImpl) Parse($$lex $$Lexer) int {
 	var $$n int
-	var $$lval $$SymType
```
Comment (tamird) [inline yacc.go]: ditto this change; why make this a member if it's never reused?

File: yacc.go (hunk 2 above)
Comment (petermattis) [inline yacc.go]: `rcvr.stack` is allocated with the `rcvr`, not a separate allocation. Ditto for `rcvr.lval`.

Comment (tamird) [inline yacc.go]: Ah, that's a good reason. Thanks for clarifying.

File: yacc.go (hunk 3 above)
Comment (tamird) [inline yacc.go]: > `rcvr.stack` is allocated with the `rcvr`, not a separate allocation. Ditto for `rcvr.lval`

Wasn't this variable previously on stack, though? And if it's a pointer, isn't it still being allocated separately?

Comment (petermattis) [inline yacc.go]: The address is passed to the lexer so escape analysis puts it on the heap. Embedding it in `ParserImpl` ensures it is allocated with the parser and the stack and not separately.

Comment (tamird) [inline yacc.go]: Ah, I see. I guess the escape analysis is not as sophisticated as I thought (the lexer is a stateless function). Thanks again.

Comment (petermattis) [inline yacc.go]: Eh? It passes `lval` to our lexer. I can't see how escape analysis could detect the value doesn't eventually escape here.

Comment (tamird) [inline yacc.go]: Right again. I only looked at `$$lex1`, not its implementation.

Comment (petermattis) [inline yacc.go]: FYI, these changes were driven by the memory profiler. It will show you the line allocations are occurring on so you don't have to guess about whether escape analysis is placing something on the stack or on the heap.
